### PR TITLE
Allow masks in select expressions

### LINKED ISSUE #2816: Add support for don't care element in tuple for parser select
```p4
state ip {
        p.extract(headers.ipv4);
	transition select(headers.ipv4.protocol, headers.ipv4.diffserv) {
	(_, _) : accept;
        default : reject;
    }
```

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Fixes #2816 
@osinstom : you may want to consider applying similar fixes to the ubpf back-end.
In fact, this scheme is better than it used to be, but it is still not general enough: it won't handle select expressions where the expression has more than 64 bits combined.

### CODE REVIEW

**Comment by hesingh** [comment]
Not a surprise that 37 ebpf tests have failed.  The reference output for these tests have to be updated.

**Comment by hesingh** [comment]
The same error is seen for  few tests I checked.
```
*** Error 2: Failed to compile P4:
/p4c/testdata/p4_16_samples/key-issue-1020_ebpf.p4(16): [--Werror=unsupported] error: DefaultExpression: not yet implemented
            default : reject;
            ^^^^^^^
make: *** [/p4c/build/tmpv41__tqz/test.c] Error 1
```

**Comment by hesingh** [comment]
```
285/1982 Test #658: ubpf/testdata/p4_16_samples/lookahead_ubpf.p4 .................................***Failed    0.32 sec
*** Error 2: Failed to compile P4:
/p4c/testdata/p4_16_samples/lookahead_ubpf.p4(37): [--Werror=unsupported] error: &&&: not yet implemented
            default: accept;
            ^^^^^^^
make: *** [/p4c/build/tmpiggk360_/test.c] Error 1
```

**Comment by mbudiu-vmw** [comment]
@osinstom : I had to modify the ubpf back-end in a similar way with the ebpf back-end. 
There is a lot of code duplication between the two, it would have been nice to reuse it, but given the actual code structure at this point it would be a lot of work to unify them. So I had to essentially make the same changes in your back-end.
Can you please review these? 
The changes strictly increase the range of programs accepted: they can now handle selects with masks in parsers.
Note that default is now converted to a mask of all 0s in the SingleArgumentSelect pass, that's why I had to modify the ubpf backend as well.

**Comment by hesingh** [comment]
The code changes need more work.

Only if an if-statement in generated C code includes an equality check with a mask, then double parenthesis are needed, e.g.,

`if ((select_7 & 0x0) == (0x0 & 0x0))`

However, if the generated code is this, the double parenthesis generated code is wrong and causes a C compiler error. 

`if ((select_5 == 65359))`

```
error: equality comparison with extraneous parentheses [-Werror,-Wparentheses-equality]
        if ((select_5 == 65359))goto parse_rlc;
             ~~~~~~~~~^~~~~~~~
```

**Comment by hesingh** [comment]
The double parenthesis are only needed in generated code if the tuple includes any don't care, such as in the P4 code example below.

```
state parse_gtpu {
	pkt.extract(hdr.gtpu_hdr);
	transition select(hdr.gtpu_hdr.e, hdr.gtpu_hdr.s, hdr.gtpu_hdr.pn) {
            (_, _, _) : parse_gtpopt;
	    (_, 1,_) : parse_gtpopt;
	    (_, _, 1) : parse_gtpopt;
	    default   : parse_ue_hdr;
	}
    }
```

because this is the generated code

`if ((select_7 & 0x0) == (0x0 & 0x0))goto parse_gtpopt;`

**Comment by hesingh** [comment]
This code fixes the issue I just reported today.

```
bool StateTranslationVisitor::preorder(const IR::SelectCase* selectCase) {
    builder->emitIndent();
    if (auto mask = selectCase->keyset->to<IR::Mask>()) {
        builder->appendFormat("if ((%s", selectValue);
        builder->append(" & ");
        visit(mask->right);
        builder->append(") == (");
        visit(mask->left);
        builder->append(" & ");
        visit(mask->right);
	builder->append("))");
    } else {
        builder->appendFormat("if (%s", selectValue);
        builder->append(" == ");
        visit(selectCase->keyset);
	 builder->append(")");
    }
    builder->append(" goto ");
```

**Comment by hesingh** [comment]
This PR lacks a test program to test don't care in a tuple used by parser select.  I have attached a P4 program.

[ebpf-tuple-p4.txt](https://github.com/p4lang/p4c/files/6795373/ebpf-tuple-p4.txt)

However, when I use the program with this PR code changes, I hit a `Compiler Bug`.  Why should  20 bits not be allowed in the tuple that `widthCheck` in `p4c/backends/ebpf/codeGen.cpp` fails.

```$ ./p4c-ebpf ../testdata/p4_16_samples/ebpf-tuple.p4 -o foo.c
In file: /home/hemant/p4c-clone/backends/ebpf/codeGen.cpp:395
Compiler Bug: ../testdata/p4_16_samples/ebpf-tuple.p4(43): (bit<20>)((bit<12>)headers.ipv4.ihl << 8 | (bit<12>)headers.ipv4.protocol & 255) << 8 | (bit<20>)headers.ipv4.diffserv & 255;: Computations on 20 bits not yet supported
 transition select(headers.ipv4.ihl, headers.ipv4.protocol, headers.ipv4.diffserv) {
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by hesingh** [comment]
@mbudiu-vmw Please expedite this PR when you get a chance.  Also, do we allow non-multiple of 8 bits for sum of all elements in he tuple? thanks.

File: midend/singleArgumentSelect.cpp [hunk 1]
```diff
@@ -2,29 +2,67 @@
...
+        expr = m->left;
+        mask = m->right;
+        hasMask = true;
+    } else if (e->is<IR::DefaultExpression>()) {
+        expr = new IR::Constant(srcInfo, type, 0, 16);
+        mask = new IR::Constant(srcInfo, type, 0, 16);
+        hasMask = true;
+    } else {
+        expr = e;
+        mask = new IR::Constant(srcInfo, type, Util::maskFromSlice(type->width_bits()-1, 0), 16);
```
Comment (fruffy) [inline midend/singleArgumentSelect.cpp:17]: What happens with zero-width bits here?

**Review by fruffy — APPROVED** [review]

File: midend/singleArgumentSelect.cpp (hunk 1 above)
Comment (mbudiu-vmw) [reply midend/singleArgumentSelect.cpp:17]: Good point. I will add a pass to eliminate operations on zero width values.

**Pull request merged**